cylc/uiserver/graphql/tornado_ws.py

```diff
@@ -0,0 +1,381 @@
+# This Cylc solution for graphql subscriptions over websockets is essentially
+# a condense from of the stagnant graphql-ws project:
+# https://github.com/graphql-python/graphql-ws
+# and our implementation of the tornado integration PR that never got merged:
+# https://github.com/graphql-python/graphql-ws/pull/25/files
+
+import asyncio
+from asyncio.queues import QueueEmpty
+from contextlib import suppress
+from inspect import isawaitable
+import json
+from weakref import WeakSet
+
+from tornado.websocket import WebSocketClosedError
+from graphql import (
+    parse,
+    validate,
+    ExecutionResult,
+    GraphQLError,
+    MiddlewareManager,
+)
+from graphql.pyutils import is_awaitable
+
+from cylc.flow.network.graphql import (
+    execution_result_to_dict,
+    instantiate_middleware,
+)
+
+from cylc.uiserver.authorise import AuthorizationMiddleware
+from cylc.uiserver.graphql.subscribe import subscribe
+
+
+NO_MSG_DELAY = 1.0
+
+GRAPHQL_WS = "graphql-ws"
```

Comment:
@oliver-sanders please check the websocket subprotocol.

Reply:
What is needed to be done in order to switch to the `graphql-transport-ws` subprotocol? (which I think is described here: https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverWebSocket.md)

Switching to `graphql-transport-ws` at some point is needed to address https://github.com/cylc/cylc-ui/issues/1028

Reply:
Actually, I was curious and had a little dig into it. Seems like this module can be lightly modified to support the `graphql-transport-ws` subprotocol, although ideally we would go with a pre-rolled solution if one is available?

Reply:
Well, the communication structure already fits the definition (aside from close events maybe?), with channels and ids etc..

Reply:
> ideally we would go with a pre-rolled solution if one is available?

There is no off-the-shelf solution for websocket support in Tornado, hence this module.

The diff required to support the graphql-transport-ws subprotocol is fairly minimal. There's an (unmerged) example in the abandoned graphql-ws library: https://github.com/graphql-python/graphql-ws/pull/65

I would suggest sticking with the current subprotocol (as David has implemented) for this graphql-core v3 upgrade. We can tackle the subprotocol change later, no need to do it now.

Reply:
Have checked the subprotocol, it appears to be producing identical results and is following the correct subprotocol for the current UI implmentation.